sklearn/utils/estimator_checks.py

```diff
@@ -2433,7 +2433,7 @@ def check_estimators_overwrite_params(name, estimator_orig):
 @ignore_warnings(category=FutureWarning)
 def check_no_attributes_set_in_init(name, estimator_orig):
     """Check setting during init. """
-    estimator = clone(estimator_orig)
+    estimator = _construct_instance(estimator_orig.__class__)
```

Comment:
The only issue with this solution is that it will fail on estimators which have non-trivial `required_param` set, otherwise should work. WDYT @amueller ?

Reply:
indeed. I don't think they are tested now but someone could construct such an estimator and pass it to ``check_estimator``. Why is this change necessary?

Reply:
This PR makes `get_params` fail if there's a required parameter which is not available through `getattr(self, attr)`, which means then `clone` itself would fail.

The issue is some estimators which would fail at implementing the part of the API tested by `check_estimators_overwrite_params` would then also fail to `clone`, and therefore this test can't really do what it's trying to do.

We're trying to test for `check_estimators_overwrite_params` here and give the user the right error message, but now the `clone` itself fails.

It can be that now `clone` is much more strict with the change in this PR, and therefore this test is not needed?

Reply:
I'm not sure I understand why `clone` is an issue in this check but not in the other checks?

Reply:
> I'm not sure I understand why `clone` is an issue in this check but not in the other checks?

Let me explain carefully. The ``check_no_attributes_set_in_init`` check is used to ensure that no attributes (ending with a trailing underscore) are set apart from the (hyper)parameters during ``__init__`` and all (hyper)parameters are stored during ``__init__``.

The issue arises because if any of these non-conformant problems happens, then, the ``get_params`` method raises a ``ValueError`` (``"The parameter cannot be retrieved as an instance attribute"``) before this check can be carried out.

The ``clone`` function is an issue only for this check because the ``test_check_no_attributes_set_in_init`` test is the only one wrongly setting the attributes and (hyper)parameters.

This will happen for any other checks calling ``clone`` if the given estimator does not fulfil the previous requirements (first paragraph). However, I do not think that this is a major issue in other checks because the ``check_no_attributes_set_in_init`` function is the first check in ``_yield_checks``.